### ISSUE TITLE
Load reports asynchronously

### ISSUE THREAD

**Issue by colinxfleming** [DCAFEngineering/dcaf_case_management#1095]
Thanks for creating an issue! Please fill out this form so we can be sure to have all the information we need, and to minimize back and forth.

* **What are we trying to do?**
The thing preventing us from using the reports endpoint right now is the page load timing out because of the heavy db load. One way around this is to load the reports asynchronously. We should do that!

* **What feature or behavior is this required for?**
Actually using reports.

* **How could we solve this issue? (Not knowing is okay!)**
I forget the most rails-y thing to do here is -- it might be setting up individual report endpoints and set up the individual partials to load with a `data-load`. Taker's choice!

* **Anything else?**
See also #1009

**Comment by tingaloo** [comment]
What do you think of having one chart displaying at a time? initially load the weekly patients data first, then having the rest as toggle options which dynamically change the graph?

**Comment by colinxfleming** [comment]
@tingaloo I think I'd prefer it as is (I think the ability to compare weeks/months/ytd was a requested feature by the CM brass) but I'm open to another solution here. I think long term, we were going to explore having these more static reports display on page load, and then have a more dynamic/d3-or-tableau-esque thing like keshif on another tab.

What do you have in mind here? Interested in your experience and insight here for sure.

**Comment by colinxfleming** [comment]
This looks like a particularly good solution for this -- https://semaphoreci.com/blog/2017/06/08/speeding-up-rails-pages-with-render-async.html

**Comment by tingaloo** [comment]
I see that the DB loads are initially done in the index with ```@weekly_report,@monthly_report,@yearly_report```.

Is this 27 DB calls at most? I was thinking of making it a background job with https://github.com/brandonhilkert/sucker_punch. Otherwise putting the reports in individual endpoints and using the render_async gem should work.

**Comment by colinxfleming** [comment]
I think an individual endpoint + render_async is the optimal way to go here. I love sucker punch and it's built with heroku dynos in mind, but it's at its best when you can just fire off something and not need a result to come back to your view (like emails, kicking off an ETL job, stuff on that order). render_async looks more akin to resolving promises (and looks way, way simpler to implement than sucker punch, tbh).

**Comment by tingaloo** [comment]
Gave this a shot, adding ```render_async``` automatically throws console errors like this.
```Refused to execute inline script because it violates the following Content Security Policy directive: "script-src 'self' 'unsafe-eval'". Either the 'unsafe-inline' keyword, a hash ('sha256-bYIgmUdITYFyqsXboR6JFTvMNyGAMxG1UuiDWGjn5ug='), or a nonce ('nonce-...') is required to enable inline execution.```
Is this something particular to the environment?
Here is what I've got so far. 
https://github.com/tingaloo/dcaf_case_management/tree/frontend/1095

**Comment by colinxfleming** [comment]
Hop on slack and we can talk thru this. Those errors are from our CSP, but that should just be set to warn rather than stop.

**Comment by colinxfleming** [comment]
Despite working in dev this is still fighting back in prod -- nonce shit isn't blending right. Investigating why, reopening in the meantime.

**Comment by colinxfleming** [comment]
Note to myself to unscrew the CSPs here and clear some runway for @lapolonio . I'd expect to see some movement on that this weekend.

Relatedly, it looks like https://github.com/ankane/chartkick is nonce-compatible, so it'll blend with our CSP, which is good news! So I think converting over what we have now to chartkick (see `app/views/reports/_patient_report.html.erb`, particularly the ugmo js block) is a good thing that we'll want to get started with. (Again, we can temporarily 'hack' around the strict CSP by heading into `config/initializers/csp.rb` and removing `unsafe-inline` from script-src.)

**Comment by tingaloo** [comment]
I have no issues with ```unsafe-inline```. Why do we need the CSP file? Does Rails lack JS security?

**Comment by lapolonio** [comment]
Appearantly
> Note that 'unsafe-inline' is ignored if either a hash or nonce value is present in the source list.

so I think the csp script_src line should be:
`script_src: %w('self' 'unsafe-eval' 'unsafe-inline'),`
and removal of nonce from render_async

Given these suggestions are what is used. It seems that we don't need chartkick and can keep c3?

**Comment by colinxfleming** [comment]
@tingaloo @lapolonio short version: I'm good with either c3 or chartkick, we should go with preference on this. That said I'd really like to keep `unsafe-inline` disallowed.

**Longwinded posting about CSPs**: Our CSP was the result of a security finding during our initial review. Rails does a bunch of good security stuff out of the box (see the security doc here http://guides.rubyonrails.org/security.html ), and the CSP functions as a supplement to that -- it winds up nearly removing XSS as an attack vector entirely. (It also kinda serves as a de facto second layer of js safety when making changes, which I sort of appreciate). In short, rails tackles 90% of 'what is that, exactly?' type app security for us, and the last 10% we're covering with a CSP, brakeman, sqreen, bundler audit, and ruby audit.

`unsafe-eval` needs to stick around because of jquery, but turning on `unsafe-inline` is a REALLY tough sell for me, because it sorta does open up an attack vector.

Some quick-ish reads on this from github (https://github.com/blog/1477-content-security-policy) and sqreen (https://blog.sqreen.io/content-security-policy/ , which has a pretty interesting link to a different github page on their experience).

**Comment by tingaloo** [comment]
Hm, if inline JS is out. I'm thinking we include a ```patient_report.js.erb``` file that houses the ```<%= javascript_tag do %>...``` stuff. Then render the .html partial ```patient_report.html.erb```

Unless the javascript is failing elsewhere, i think that is a potential solution.

**Comment by lapolonio** [comment]
I agree keeping `unsafe-inline` out is ideal to keep in for security reasons.

Current report pattern:
server side async loading of DOM and javascript code which is automatically executed and renders the graph

Issue:
Current CSP config doesn't allow for inline scripting. However this make it really hard for javascript to know that the query is done and async DOM is present and ready to render graph.

Proposed solution:
Javascript ajax request for query. then manages DOM manipulation and calls c3 to render graph. Most javascript frameworks can bind to a DOM and manipulate. I recommend choosing a light framework like knockout.js vs angular or react which are full 'heavy' javascript solutions. I've worked with knockout before vue.js looked cool.

Caveat:
As long as we don't need to manage application state between routes, like remember which report was open when we navigate back from another page. this should be fine. otherwise we would also need to manage application state and routing in javascript.

**Comment by colinxfleming** [comment]
Vue looks pretty lightweight, might be a good solution here if we do need to adopt a js-first approach to this. (We def don't need to manage app state between routes.)

My inclination is to make a last weekend push with the existing reports code (render_async partials with nonces, c3 or chartkick), before we put a js framework in place. @lapolonio can we poke at this together tonight/tomorrow and see if we can come up with something? I'll be sure to post a stream of consciousness log in slack when I'm jabbing at this.

**Comment by lapolonio** [comment]
Sure.

Note there is a difference between turbolinks navigate vs page refresh. On page refresh the entire page is loaded with DOM and script for each graph. With turbolinks navigation the async loaded script is blocked by CSP.

**Comment by colinxfleming** [comment]
@tingaloo had a great comment in #1183 that I wanted to expand on -- apologies to you both for not saying this sooner! We shouldn't feel married to c3 if there's agreement among us that chartkick would be easier to implement and maintain. The c3 code got built on the fly as part of a hackathon to display the results of the reporting code, which currently lives in `app/lib/*`, and c3 was just what that dev felt most comfortable with on the fly. So we should 100% feel entitled to tear that out and replace it with something else, especially since our reporting work has been so speculative to this point.

In short, if we're in agreement that we should move to chartkick, let's go for it. @tingaloo and @lapolonio want to weigh in?

**Comment by colinxfleming** [comment]
@lapolonio  the turbolinks thing is is really interesting and a great flag -- the behavior I'm seeing on that front is that the reports load fine the first time and then choke if you navigate away, add a contact or patient, and then return to the reports view. Is that about what you're seeing?

**Comment by tingaloo** [comment]
I heard good things about chartkick. I'm all about upgrading over
maintaining, so go for it!

**Comment by lapolonio** [comment]
@colinxfleming i haven't seen that but I can try to reproduce locally

I guess I can try to port the graph code over to chartkick

**Comment by colinxfleming** [comment]
@lapolonio Yeah, if you could port the chart code over to chartkick that would be lovely.

**Comment by lapolonio** [comment]
@colinxfleming i started incorporating chartkick. one thing i can't figure out is how to incorporate nonce to the generated javascript

There seems to be a discrepancy between https://github.com/ankane/chartkick/blob/f42f3bb732c9f7a9871da31a57935bfc3f1a2c35/lib/chartkick/helper.rb and local installed from gem helper.rb (which doesn't contain the nonce code)

**Comment by colinxfleming** [comment]
@lapolonio do me a solid and push up a work in progress pull request?

It looks like the nonce handling is at https://github.com/ankane/chartkick/blob/f42f3bb732c9f7a9871da31a57935bfc3f1a2c35/lib/chartkick/helper.rb#L49 . You can see what's in the gem you actually have by doing `gem open chartkick` and nav-ing to the right spot. Any weird discrepancies there?

**Comment by lapolonio** [comment]
@colinxfleming o i installed an old version (1.2.4) i updated (2.2.4)
another note the json data has to change to get the same column chart
from
```
{
 "DC": {"name":"DC", "patients_contacted":2, "new_patients_contacted":2, "pledges_sent":0},
 "MD": {"name":"MD", "patients_contacted":0, "new_patients_contacted":0, "pledges_sent":0},
 "VA": {"name":"VA", "patients_contacted":0, "new_patients_contacted":0, "pledges_sent":0}
}
```
to:
```
{name: "patients_contacted", data: {"DC":2, "MD":0, "VA":0}},
{name: "new_patients_contacted", data: {"DC":2, "MD":0, "VA":0}},
{name: "pledges_sent", data: {"DC":0, "MD":0, "VA":0}}
```

which is totally different from how it is pulled in Line_Summary.
You can test by going to https://www.chartkick.com/ and executing in Inspector (Chrome)
```
data = [
    {name: "patients_contacted", data: {"DC":2, "MD":0, "VA":0}},
    {name: "new_patients_contacted", data: {"DC":2, "MD":0, "VA":0}},
    {name: "pledges_sent", data: {"DC":0, "MD":0, "VA":0}},
]
new Chartkick.ColumnChart("chart-1", data)
```

**Comment by colinxfleming** [comment]
Migrating into #634
